File: Packages/Core/Sources/FissionCore/Domains/Thread/SSHDirectoryListing.swift
import Foundation

/// Lists child directories on a Remote Machine over SSH (BatchMode, no password prompt).
public enum SSHDirectoryListing {
    public static func processArguments(
        machine: RemoteMachine,
        directory: String
    ) -> [String] {
        var arguments = [
            "-o", "BatchMode=yes",
            "-o", "ConnectTimeout=4",
            "-o", "LogLevel=ERROR"
        ]
        if let sshPort = machine.sshPort, sshPort != 22 {
            arguments.append(contentsOf: ["-p", String(sshPort)])
        }
        arguments.append(machine.target)
        arguments.append(remoteCommand(directory: directory))
        return arguments
    }

    public static func remoteCommand(directory: String) -> String {
        "sh -c \(MoshCommand.posixQuote(listingScript)) -- \(MoshCommand.posixQuote(directory))"
    }

    public static func parseOutput(_ output: String) -> [String] {
        var seen = Set<String>()
        var names: [String] = []
        for line in output.split(whereSeparator: \.isNewline) {
            let name = String(line)
            guard !name.isEmpty, name != ".", name != "..", !name.hasPrefix(".") else { continue }
            guard seen.insert(name).inserted else { continue }
            names.append(name)
        }
        return names.sorted {
            $0.localizedCaseInsensitiveCompare($1) == .orderedAscending
        }
    }

    public static func parseFixtureJSON(_ json: String) -> [String: [String]]? {
        guard let data = json.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode([String: [String]].self, from: data)
    }

    private static let listingScript = """
    path=$1
    case "$path" in
    ~) dir="$HOME" ;;
    ~/*) dir="$HOME/${path#~/}" ;;
    *) dir="$path" ;;
    esac
    [ -d "$dir" ] || exit 0
    find "$dir" -mindepth 1 -maxdepth 1 ! -name '.*' -print 2>/dev/null | while IFS= read -r p; do
      [ -d "$p" ] || continue
      printf '%s\\n' "${p##*/}"
    done
    """
}

File: packages/electron/src/shared/sshDirectoryListing.ts
import { posixQuote } from "./moshCommand";
import { isValidMachine, machineTarget } from "./remoteMachine";
import type { RemoteMachine } from "./types";

export type RemoteDirectoryListing =
  | { kind: "missing" }
  | { kind: "failed" }
  | { kind: "contents"; names: string[] };

export const notADirectoryStatus = 2;

const listingScript = `path=$1
case "$path" in
~) dir="$HOME" ;;
~/*) dir="$HOME/\${path#~/}" ;;
*) dir="$path" ;;
esac
[ -d "$dir" ] || exit 2
find "$dir" -mindepth 1 -maxdepth 1 ! -name '.*' -print 2>/dev/null | while IFS= read -r p; do
  [ -d "$p" ] || continue
  printf '%s\\n' "\${p##*/}"
done`;

export function remoteCommand(directory: string): string {
  return `sh -c ${posixQuote(listingScript)} -- ${posixQuote(directory)}`;
}

export function processArguments(machine: RemoteMachine, directory: string): string[] {
  const arguments_ = [
    "-o",
    "BatchMode=yes",
    "-o",
    "ConnectTimeout=4",
    "-o",
    "LogLevel=ERROR"
  ];
  if (machine.sshPort != null && machine.sshPort !== 22) {
    arguments_.push("-p", String(machine.sshPort));
  }
  arguments_.push(machineTarget(machine), remoteCommand(directory));
  return arguments_;
}

export function parseOutput(output: string): string[] {
  const seen = new Set<string>();
  const names: string[] = [];
  for (const line of output.split(/\r?\n/)) {
    const name = line;
    if (name.length === 0 || name === "." || name === ".." || name.startsWith(".")) {
      continue;
    }
    if (seen.has(name)) {
      continue;
    }
    seen.add(name);
    names.push(name);
  }
  return names.sort((left, right) =>
    left.localeCompare(right, undefined, { sensitivity: "accent" })
  );
}

export function parseFixtureJSON(json: string): Record<string, string[]> | null {
  try {
    const parsed: unknown = JSON.parse(json);
    if (parsed == null || typeof parsed !== "object" || Array.isArray(parsed)) {
      return null;
    }
    const result: Record<string, string[]> = {};
    for (const [key, value] of Object.entries(parsed)) {
      if (!Array.isArray(value) || value.some((item) => typeof item !== "string")) {
        return null;
      }
      result[key] = value;
    }
    return result;
  } catch {
    return null;
  }
}

export function listingResult(status: number, output: string): RemoteDirectoryListing {
  if (status === 0) {
    return { kind: "contents", names: parseOutput(output) };
  }
  if (status === notADirectoryStatus) {
    return { kind: "missing" };
  }
  return { kind: "failed" };
}

export function canList(machine: RemoteMachine): boolean {
  return isValidMachine(machine);
}
